api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -337,16 +337,22 @@ private void checkAndAddPartitionName(String name) {
     }
 
     private void checkAndAddPartitionName(String name, Integer sourceColumnId) {
+      checkAndAddPartitionName(name, sourceColumnId, true);
+    }
+
+    private void checkAndAddPartitionName(String name, Integer sourceColumnId, boolean checkConflict) {
       Types.NestedField schemaField = schema.findField(name);
-      if (sourceColumnId != null) {
-        // for identity transform case we allow  conflicts between partition and schema field name as
-        //   long as they are sourced from the same schema field
-        Preconditions.checkArgument(schemaField == null || schemaField.fieldId() == sourceColumnId,
-            "Cannot create identity partition sourced from different field in schema: %s", name);
-      } else {
-        // for all other transforms we don't allow conflicts between partition name and schema field name
-        Preconditions.checkArgument(schemaField == null,
-            "Cannot create partition from name that exists in schema: %s", name);
+      if (checkConflict) {
```

Comment:
I wonder if `checkConflict` should just be a part of the builder? It seems a little odd that just the identity method has this flag. Like maybe you have something like

partitionSpec.builder().checkConflicts(false) ?

Reply:
I do agree that it might make sense to add to the builder. However, since it's a pretty specific occurrence (and really only applies to `identity`) would that be possibly confusing to users about when it should be called and what it would be applied to?

It seems that `BaseMetadataTable#transformSpec` is the only place that uses `identity` with this boolean argument as false. Is it possible to get rid of that and then `checkAndAddPartitionName` could just have some sort of check for whether or not this is `identity` transform?

Reply:
I agree it's a little odd that just the identity method has the `checkConflict` flag. But the only place that we don't need to check conflict and allow conflict between partition name and schema field name is `BaseMetadataTable#transformSpec`. All other transforms don't allow conflicts between partition name and schema field so the `checkConflict` flag defaults to true. Since it only can be false in `BaseMetadataTable#transformSpec`, I guess it might be ok to do it in the identity method than add to the builder?

I tried to have a different function but couldn't come with a neat one. There are quite bit some code duplication. 

 I added some comment to explain a the flag. Hope it looks a little better now.

Reply:
I'm not sure the comment makes it more clear for me. I think it may be better to just remove the comment. I think I'd still rather it was a builder wide flag rather than an internal method with a flag but If you think it is cleaner this way that is ok.

I have a few other possible suggestions
    1. Move the Check for identity added fields to "checkCompatibility" that way you could just do a buildUnchecked  in TransformSpec to ignore the invalid identity
    2. Change transformSpec to just construct the Partition Spec without the builder

Reply:
I added `checkConflict` in Builder. It does look cleaner this way :)

Reply:
Agreed it does look cleaner in the builder!

Question for my own full understanding: Is it correct to say that we can ignore `checkConflict` when we know there is no conflict? Like in the transformSpec, we can set `checkConflict` to `false` because we're already adding `partitionPrefix`?

Reply:
In this case it's because we arent "really" making a partitioning spec, we are just using it to get the transform so that pushdown will work correctly